src/particle_filter.h

```diff
@@ -1,54 +1,52 @@
 #ifndef ED_LOCALIZATION_PARTICLE_FILTER_H_
 #define ED_LOCALIZATION_PARTICLE_FILTER_H_
 
+#include "kdtree.h"
+#include "transform.h"
+
 #include <geolib/datatypes.h>
 
+#include <tue/config/configuration.h>
+
 #include <vector>
 
 // ----------------------------------------------------------------------------------------------------
 
-class Transform
+struct Sample
 {
+    Sample() {}
 
-public:
-
-    inline const geo::Transform2& matrix() const { return pose_; }
-    inline double rotation() const { return rotation_; }
-    inline const geo::Vec2& translation() const { return pose_.t; }
-
-    inline void setTranslation(const geo::Vec2& v) { pose_.t = v; }
-    inline void set(const geo::Transform2& p)
-    {
-        pose_ = p;
-        rotation_ = p.rotation();
-    }
-
-    inline void setRotation(double rot)
+    Sample(const geo::Transform2& t)
     {
-        pose_.setRotation(rot);
-        rotation_ = rot;
+        pose.set(t);
     }
 
-private:
-
-    geo::Transform2 pose_;
-    double rotation_;
-
+    double weight;
+    Transform pose;
 };
 
 // ----------------------------------------------------------------------------------------------------
 
-struct Sample
+struct Cluster
 {
-    Sample() {}
-
-    Sample(const geo::Transform2& t)
+    Cluster() : count(0), weight(0), cov(0.f), m({0, 0, 0, 0}), c({ {{0, 0}, {0, 0}} })
     {
-        pose.set(t);
+        mean.set(geo::Transform2::identity());
     }
 
+    // Number of samples
+    unsigned int count;
+
+    // Total weight of samples in this cluster
     double weight;
-    Transform pose;
+
+    // Cluster statistics
+    Transform mean;
+    geo::Mat3 cov;
+
+    // Workspace
+    std::array<double, 4> m;
```

Comment:
Why these magic member variables? Could you give them some meaningful name? Or did you get these from a reference publication (and did you cite this?)

Reply:
Just from the ROS AMCL